public class PrimArrayNewError {
    // :: error: incompatible types: int[] cannot be converted to int
    int f = new int[10];
    // :: error: incompatible types: int[] cannot be converted to int
    int g = new int[20];
}
